pkg/handler/internal/default_pruning.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package internal
+
+import "github.com/go-openapi/spec"
+
+// PruneDefaults remove all the defaults recursively from all the
+// schemas in the definitions, and does not modify the definitions in
+// place.
+func PruneDefaults(definitions spec.Definitions) spec.Definitions {
+	if definitions == nil {
+		return nil
+	}
+	newDefinitions := spec.Definitions{}
+
+	for name, definition := range definitions {
+		newDefinitions[name] = *prune_schema_defaults(&definition)
+	}
+
+	return newDefinitions
+}
+
+func prune_schemalist_defaults(schemas []spec.Schema) []spec.Schema {
```

Comment:
I would prefer to see some on-demand allocation of a new list, i.e. turn all the signatures here into `func prune_*(*Foo) (bool, *Foo)` with the bool false if nothing changed.

Reply:
I've tried to come-up with a reasonable algorithm here, but that seems complicated compared to the possible benefits.

The problem is that we don't know if we need to allocate the list/map until we know if something has changed in the list. And we can't check the list and then do it without causing squared complexity over the depth of the tree.

AFAICT, we'd have to initially assume that none are going to change, and if we find one, allocate the list, copy the items we've already seen and continue with that new list, which again, seems a little overkill here.

But maybe I'm missing something?

Reply:
Compare https://github.com/kubernetes/kube-openapi/blob/d8b685cd2daadc40121e864b5e87ca8ededd9eed/pkg/aggregator/mutating_walker.go#L33. Something similar for defaults.